### ISSUE TITLE
draw sprixels bottom-to-top

### ISSUE THREAD

**Issue by joseluis** [dankamongmen/notcurses#1575]
I've created the example [`issue-1573`](https://github.com/dankamongmen/notcurses/blob/master/rust/examples/issue-1573.rs#L60), where 2 ncvisuals are being rendered with NCBLIT_PIXEL blitter. White is rendered first, and the red one is rendered last.

The bug appears when notcurses_render() is not called after the first visual render, but only after both ncvisuals has been rendered.

In that case, and only if the terminal supports pixels, then the second plane appears under the first plane.

![pixel-norender](https://user-images.githubusercontent.com/5832068/115634944-ead8a880-a30a-11eb-8f48-f839b11776ce.png)

If a call to notcurses_render is made before the second ncvisual render, then they show in the correct order:

![pixel-render](https://user-images.githubusercontent.com/5832068/115635358-d21cc280-a30b-11eb-9fd0-65a4d450c790.png)

Finally, in a terminal without pixel support they appear in the correct order under all circumstances. 

![nopixel_without-middle-render](https://user-images.githubusercontent.com/5832068/115634950-eca26c00-a30a-11eb-9391-e174a36b985a.png)

**Comment by joseluis** [comment]
Maybe related to #1501

**Comment by joseluis** [comment]
BTW I didn't expected that to render in a terminal without pixel support. I see the degradation scale in `blit.c lookup_blitset() `, which I just documented in the rust bindings…

**Comment by dankamongmen** [comment]
> BTW I didn't expected that to render in a terminal without pixel support. I see the degradation scale in `blit.c lookup_blitset() `, which I just documented in the rust bindings…

yeah if you don't want degradation you need to supply `NCVISUAL_OPTION_NODEGRADE`.

**Comment by joseluis** [comment]
BTW

> without quadrant support, NCBLIT_2x2 decays to NCBLIT_2x1

why there's no functions to check support for these?

**Comment by joseluis** [comment]
Hmmm I guess that's what `media_defblitter` is for.....

**Comment by joseluis** [comment]
Maybe related to #1462

**Comment by dankamongmen** [comment]
> BTW
> 
> > without quadrant support, NCBLIT_2x2 decays to NCBLIT_2x1
> 
> why there's no functions to check support for these?

i could certainly add them. they were unconditional up until two weeks ago or so. but yeah, i'd prefer people use other methods to determine what's available, or better yet not need to. but i could add them, no problem. especially since i expose them in `notcurses_debug_caps()` and absolutely do not want people parsing that output.

**Comment by dankamongmen** [comment]
Right now multiple sprixels don't honor one another's z-order. That needs be addressed in general, see #1501. I have it scheduled for 3.0.0 right now, but maybe it needs be moved back to 2.3.0? Closing this as a duplicate of #1501, which I'll move back to 2.3.0 for the moment. I need think about it.

**Comment by dankamongmen** [comment]
In the meantime, I added `notcurses_canquadrant()` and `notcurses_canhalfblock()`.

**Comment by dankamongmen** [comment]
so we can best get that done, i think, by (1) putting new planes on the end of the list, as opposed to the front (to match how new planes are deployed), and then (2) when a sprixel plane is moved along the z-axis,

**Comment by dankamongmen** [comment]
I've gone ahead and broken the sprixel list up into piles. I will go ahead and doubly link the list for quick removal during the rendering descent.

**Comment by dankamongmen** [comment]
I've got a first try at this ready for testing. @joseluis could you take a look at `dankamongmen/ordered-sprixels`? It definitely has at least one problem (running `notcurses-demo` in kitty never cleans up the `yield` background, which remains visible for the duration of the demo), but it will hopefully solve your problem.

**Comment by joseluis** [comment]
> In the meantime, I added `notcurses_canquadrant()` and `notcurses_canhalfblock()`.

This makes me think, wouldn't it mak sense to be able to check those and the rest of the capabilities from the direct mode? Specially if ncvisuals are going to be usable there.

**Comment by dankamongmen** [comment]
> > In the meantime, I added `notcurses_canquadrant()` and `notcurses_canhalfblock()`.
> 
> This makes me think, wouldn't it mak sense to be able to check those and the rest of the capabilities from the direct mode? Specially if ncvisuals are going to be usable there.

yeah i thought the same thing earlier and just didn't care to fuck with it at that time.

though really, i don't want people to think about this kind of thing if they can avoid it.

**Comment by joseluis** [comment]
> though really, i don't want people to think about this kind of thing if they can avoid it.

Well... as long as it's not forced upon them to use it if they don't want to... But being able to check them in full mode and not in direct mode when both employ them for displaying images... feels like something is lacking.

Although this is making me think about the fundamental difference between both modes. Isn't directmode like full mode without an altscreen? Is direct mode really needed?

**Comment by dankamongmen** [comment]
> > though really, i don't want people to think about this kind of thing if they can avoid it.
> 
> Well... as long as it's not forced upon them to use it if they don't want to... But being able to check them in full mode and not in direct mode when both employ them for displaying images... feels like something is lacking.
> 
> Although this is making me think about the fundamental difference between both modes. Isn't directmode like full mode without an altscreen? Is direct mode really needed?

you've gotta have direct mode for CLI-style programs. you can do it with rendered mode, but it's more complicated than it needs be. you can do rendered mode (what i'm calling full mode now) without the alternate screen; the difference is whether you're doing line-based output (that can freely intermix with other output) or frame-based output where you render the frame, and no other output can be involved.

**Comment by joseluis** [comment]
> I've got a first try at this ready for testing. @joseluis could you take a look at `dankamongmen/ordered-sprixels`? It definitely has at least one problem (running `notcurses-demo` in kitty never cleans up the `yield` background, which remains visible for the duration of the demo), but it will hopefully solve your problem.

It does solve the bug I encountered, but it requires a couple of changes in the merge & merge_simple NcPlane methods, that now requires mutable references to the plane. I can do it after you commit the changes.

**Comment by dankamongmen** [comment]
> It does solve the bug I encountered

awesome, thanks for verifying

> but it requires a couple of changes in the merge & merge_simple NcPlane methods, that now requires mutable references to the plane.

that's the cost of progress! =] i need to test this some more, and like i said, i know it has at least one bug. but yeah this is the solution for the problem you saw.

**Comment by dankamongmen** [comment]
another thing we need make sure we do is redraw the affected sprixels on a relative height change, even if they haven't been invalidated. probably just invalidate them when the change is detected.

**Comment by dankamongmen** [comment]
i think i might have this working now.

**Comment by dankamongmen** [comment]
yeah, gonna put this up as a PR
